src/compiler/c.rs

```diff
@@ -445,6 +561,379 @@ where
+                included_files,
+                &mut digest,
+                system,
+                config,
+                time_of_compilation,
+            )? {
+                return Ok(false);
+            };
+            // Everything of interest between hash_start and start has been hashed now.
+            hash_start = start;
+        } else if &bytes[start..start + INCBIN_DIRECTIVE.len()] == INCBIN_DIRECTIVE
+            && ((slice[7] == b' ' && (slice[8] == b'"' || (slice[8] == b'\\' && slice[9] == b'"')))
+                || slice[7] == b'"')
+        {
+            // An assembler .inc bin (without the space) statement, which could be
+            // part of inline assembly, refers to an external file. If the file
+            // changes, the hash should change as well, but finding out what file to
+            // hash is too hard for sccache, so just bail out.
+            debug!("Found potential unsupported .inc bin directive in source code");
+            return Ok(false);
+        } else if slice.starts_with(b"___________") && (start == 0 || bytes[start - 1] == b'\n') {
+            // Unfortunately the distcc-pump wrapper outputs standard output lines:
+            // __________Using distcc-pump from /usr/bin
+            // __________Using # distcc servers in pump mode
+            // __________Shutting down distcc-pump include server
+            digest.update(&bytes[hash_start..start]);
+            while start < total_len && slice[0] != b'\n' {
+                start += 1;
+                if start < total_len {
+                    slice = &bytes[start..];
+                }
+            }
+            slice = &bytes[start..];
+            if slice[0] == b'\n' {
+                start += 1;
+            }
+            hash_start = start;
+            continue;
+        } else {
+            start += 1;
+        }
+    }
+    digest.update(&bytes[hash_start..]);
+
+    Ok(true)
+}
+
+/// Copied from cargo.
+///
+/// Normalize a path, removing things like `.` and `..`.
+///
+/// CAUTION: This does not resolve symlinks (unlike
+/// [`std::fs::canonicalize`]).
+pub fn normalize_path(path: &Path) -> PathBuf {
+    use std::path::Component;
+    let mut components = path.components().peekable();
+    let mut ret = if let Some(c @ Component::Prefix(..)) = components.peek().cloned() {
+        components.next();
+        PathBuf::from(c.as_os_str())
+    } else {
+        PathBuf::new()
+    };
+
+    for component in components {
+        match component {
+            Component::Prefix(..) => unreachable!(),
+            Component::RootDir => {
+                ret.push(component.as_os_str());
+            }
+            Component::CurDir => {}
+            Component::ParentDir => {
+                ret.pop();
+            }
+            Component::Normal(c) => {
+                ret.push(c);
+            }
+        }
+    }
+    ret
+}
+
+// Returns false if the include file was "too new" and therefore should disable
```

Comment:
what means "too new" in this context ?

Reply:
It means changed at or after the start of the compilation. I'll update the docstring and move the explanation somewhere more obvious.